### DISCUSSION TITLE
Use CDC and NET device class together

### DISCUSSION THREAD

**Discussion by Wini-Buh** [hathach/tinyusb#1357] [Q&A]
In my project I'm running two different USB classes (device stack), CDC and NET. That means I can either communicate trough a serial connection (CDC class) or over a network connection via USB (RNDIS ode ECM, NET class) derived from the `net_lwip_webserver` example. So all is build depended, means I can only use one of the two classes in the project at build build.

My question: Is it possible to define a descriptor setting to let both classes use at the same time ?

Any help would be welcome

**Comment by hathach** [comment]
it is easily possible, what have you tried so far, and is the issue with your descriptor

**Comment by Wini-Buh** [comment]
I did not tried anything yet. I only use two separate descriptor sets, because I don't have any idea to merge the two sets together (in fact the descriptor settings are sometimes a little mystery to me). Maybe you could give me some advise how build this descriptor that fulfils my requirements ? 

Here is my source file containing the descriptor settings for the two classes. The macro `LWIP_ACTIV` switches between the two settings.

[usb_descriptors.zip](https://github.com/hathach/tinyusb/files/8151714/usb_descriptors.zip)

**Reply by hathach** [reply]
check out the example for multiple interfaces here (cdc + msc) https://github.com/hathach/tinyusb/blob/master/examples/device/cdc_msc/src/usb_descriptors.c#L143

**Reply by Wini-Buh** [reply]
@hathach Sorry to trouble you about it, but I think I need your help. I have checked out the example and understand it so far. The descriptors implementation for the two configurations RNDIS or ECM are in my few a little bit different.

I tried to build the "descriptor chain" for 
- RNDIS and CDC
- ECM and CDC

but without any success. That means I should see a virtual network interface (with RNDIS or ECM support) and a virtual com port on the host PC. 

Maybe you could copy & past an example of such a "descriptor chain" ?
Thank you very much for your help in advance.

**Reply by Wini-Buh** [reply]
@hathach  I got a working descriptor definition now, but it only works with a Linux host. If I try to connect to a Windows host (Win 10), it shows a unknown "USB Composite Device" in the device manager. 

With the Linux host, I can access the embedded web server over the TinyUSB stack and also open a virtual com port and enter commands on the embedded command line at the same time.

Do you have an idea what's wrong here ? I have attached the descriptor file and the TinyUSB debug log output.
[Decriptor_And_Log.zip](https://github.com/hathach/tinyusb/files/8188757/Decriptor_And_Log.zip)

**Reply by hathach** [reply]
RNDIS is a subclass of cdc, it also have 2 interfaces

```C
enum {
  ITF_NUM_CDC = 0,
  ITF_NUM_CDC_DATA,
  ITF_NUM_RNDIS,
// missing 1 interface count here. Linux is as always more forgiving than others.
  ITF_NUM_TOTAL
};
```

**Reply by Wini-Buh** [reply]
Thank you very much for your hint. I changed the enum to this 

```
enum {
  ITF_NUM_CDC = 0,
  ITF_NUM_CDC_DATA,
  ITF_NUM_RNDIS,
  ITF_NUM_RNDIS_DATA,
  ITF_NUM_TOTAL
};
```

On the device manager of Windows host, I now see the serial (USB) interface, which is working. I also see the "Remote NDIS based Internet Sharing Device" in the network adapter section but it's not working. The icon has a yellow warning overlay.

Any more hints ?

**Reply by hathach** [reply]
your descriptor look ok, to double check, please post your out as attached txt from following tool:  
- on windows, use this usbtreeview https://www.uwe-sieber.de/usbtreeview_e.html 
- on linux `lsusb -v`

you could try to 
- change the vid/pid in case windows cache your pid. also maybe
- try to have RNDIS interface before the cdc

**Reply by Wini-Buh** [reply]
Good catch, you did it... Your tip to change the order (RNDIS interface before CDC) did the trick. I never thought it would be so sensitive. It now works with a Linux and a Windows host. 

Attached you will find the output of the Windows tool you mentioned. The file `Windows_NotWorking_CDC_RNDIS` is the output with the descriptor definition, which does not work (the one I already uploaded here, with the correction of the interface counting). The file `Windows_Working_RNDIS_CDC` contains the output with the changed descriptor, which works.

I also attached two files (`lsusb_CDC_RNDIS` and `lsusb_RNDIS_CDC`) the output (lsusb command) from the Linux hosts for both descriptor variants (both are working so far on the Linux host).

I hope this gives you deeper sight, were the problem of the dependency of the ordering in the descriptor definition could be.

[lsusb_CDC_RNDIS.txt](https://github.com/hathach/tinyusb/files/8204436/lsusb_CDC_RNDIS.txt)
[lsusb_RNDIS_CDC.txt](https://github.com/hathach/tinyusb/files/8204437/lsusb_RNDIS_CDC.txt)
[Windows_NotWorking_CDC_RNDIS.txt](https://github.com/hathach/tinyusb/files/8204438/Windows_NotWorking_CDC_RNDIS.txt)
[Windows_Working_RNDIS_CDC.txt](https://github.com/hathach/tinyusb/files/8204439/Windows_Working_RNDIS_CDC.txt)

**Reply by hathach** [reply]
the problem can be the caching issue with windows if you previously testing with the same pid e.g testing the net lwip webserver. You could try to test with a new pc, or try to change the pid value or uninstall current driver for your pid.

Note: windows use vid + pid + bcd combination to cache the driver binding, it won't perform configuration parsing if the device is previously recognized and just use the cached driver for fast enumeration.

**Reply by Wini-Buh** [reply]
Made a test with a changed PID and the "CDC-RNDIS" descriptor definition. Windows installs a new driver for it, but in the device manager the RNDIS device is marked as not ready (device could not be started). So it doesn't seems a problem with the caching. Could you see anything in the previously attached data files ?

**Reply by Wini-Buh** [reply]
@hathach Sorry to trouble you again. Did you found some time to have a look into the previously attached data files ?

I also tried to connect to a mobile telephone running Android 11 using a serial app "Serial USB Terminal 1.45". It has the same behaviour like the Windows host. If the RNDIS interface is defined before the CDC in the descriptor, it is not possible to communicate over the terminal. The Serial USB Terminal shows the USB device correctly, but a communication is not possible. If the ordering is changed the serial communication is working with the Android host.

**Reply by hathach** [reply]
@Wini-Buh sorry, I am too busy with other paid work and haven't looked at this issue. Regarding the Android issue, you could try to get its kernel message to see if there is any useful info there.

Btw: you should provide an minimal example compilable in similar fashion to `net_lwip_webserver` for others and me (when having time) to reproduce it. Also you should also try to run it on other mcus to make sure it is not rx specific issue.

**Reply by Wini-Buh** [reply]
> @Wini-Buh sorry, I am too busy with other paid work and haven't looked at this issue. 

No problem at all, I understand that
> Regarding the Android issue, you could try to get its kernel message to see if there is any useful info there.

Ok, I can try that... It's not very urgent at the moment
> Btw: you should provide an minimal example compilable in similar fashion to `net_lwip_webserver` for others and me (when having time) to reproduce it. Also you should also try to run it on other mcus to make sure it is not rx specific issue.

I will expand/modify the `net_lwip_webserver` example to reproduce the problem. Unfortunately I can't try it with another MCU, because all hardware with USB, which I own is only equipped with RX controllers.
